flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/AbstractFlinkService.java

```diff
@@ -898,58 +901,54 @@ public JobDetailsInfo getJobDetailsInfo(JobID jobID, Configuration conf) throws
         }
     }
 
-    /** Wait until the FLink cluster has completely shut down. */
-    @VisibleForTesting
-    void waitForClusterShutdown(String namespace, String clusterId, long shutdownTimeout) {
-        LOG.info("Waiting for cluster shutdown...");
-
-        boolean jobManagerRunning = true;
-        boolean taskManagerRunning = true;
-        boolean serviceRunning = true;
+    /** Returns a list of Kubernetes Deployment names for given cluster. */
+    protected abstract List<String> getDeploymentNames(String namespace, String clusterId);
 
-        for (int i = 0; i < shutdownTimeout; i++) {
-            if (jobManagerRunning) {
-                PodList jmPodList = getJmPodList(namespace, clusterId);
+    /** Wait until the FLink cluster has completely shut down. */
+    protected void waitForClusterShutdown(
+            String namespace, String clusterId, long shutdownTimeout) {
+        long timeoutAt = System.currentTimeMillis() + shutdownTimeout * 1000;
+        LOG.info("Waiting {} seconds for cluster shutdown...", shutdownTimeout);
 
-                if (jmPodList == null || jmPodList.getItems().isEmpty()) {
-                    jobManagerRunning = false;
-                }
-            }
-            if (taskManagerRunning) {
-                PodList tmPodList = getTmPodList(namespace, clusterId);
+        for (var deploymentName : getDeploymentNames(namespace, clusterId)) {
+            long deploymentTimeout = timeoutAt - System.currentTimeMillis();
 
-                if (tmPodList.getItems().isEmpty()) {
-                    taskManagerRunning = false;
-                }
+            if (!waitForDeploymentToBeRemoved(namespace, deploymentName, deploymentTimeout)) {
+                LOG.error(
```

Comment:
In case of a timeout, we just log the message as an error, but during upgrade this means that the operator will just proceed to create a new `Deployment`, then get `AlreadyExists` error, resulting in a deadlock.

Not sure what else can we do in this case.

Reply:
I fully agree, but don't have a solution. At least the default value for the cleanup timeout could be increased to a higher value. What do you think?

Reply:
I don't think there could be any downside to setting the default cleanup timeout to e.g. 5 minutes. @gyfora WDYT? Should that be done in another PR or should I add that here, or do you think we should leave the timeout as is?

Reply:
If no trouble for you @mateczagany let's add the default timeout increase here. I am not a big fan of too many PRs it's just more work to review :)

Reply:
5 minutes seems reasonable for prod envs